Move nexusmutual items from v1 to v2

sharkxff opened the pull request:
Brief comments on the purpose of your changes:


*For Dune Engine V2*
I've checked that:
General checks:
* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [x] if adding a new model, I added a test
* [x] the filename is unique and ends with .sql
* [x] each sql file is a select statement and has only one view, table or function defined  
* [x] column names are `lowercase_snake_cased`
* [x] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [x] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributors)

Pricing checks:
* [ ] `coin_id` represents the ID of the coin on coinpaprika.com
* [ ] all the coins are active on coinpaprika.com (please remove inactive ones)

Join logic:
* [x] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

Incremental logic:
* [x] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [x] where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

dune-eng commented:
Workflow run id 3376473595 approved.

dune-eng commented:
Workflow run id 3376473590 approved.

sharkxff commented:
I've checked that General checks

Hosuke commented on models/nxm/ethereum/nxm_ethereum_schema.yml:
```diff
@@ -0,0 +1,103 @@
+      contributors: [zxf, jonah]
+    config:
+      tags: ['ethereum','nexusmutual','trades']
+    description: >
+        nexusmutual quotation trades on Ethereum
+    columns:
```
Hi @sharkxff , you can add an uniqueness test here.
```suggestion
    description: >
        nexusmutual quotation trades on Ethereum
    tests:
      - dbt_utils.unique_combination_of_columns:
          combination_of_columns:
            - evt_block_number
            - evt_tx_hash
            - evt_index
    columns:
```

dune-eng commented:
Workflow run id 3390877579 approved.

dune-eng commented:
Workflow run id 3390877573 approved.

dune-eng commented:
Workflow run id 3390896065 approved.

dune-eng commented:
Workflow run id 3390896063 approved.

dune-eng commented:
Workflow run id 3391025953 approved.

dune-eng commented:
Workflow run id 3391025954 approved.

dune-eng commented:
Workflow run id 3391165179 approved.

dune-eng commented:
Workflow run id 3391165169 approved.

Hosuke commented on models/nxm/ethereum/nxm_ethereum_quotation_trades.sql:
```diff
@@ -0,0 +1,79 @@
+	materialized = 'incremental',
+	file_format = 'delta',
+	incremental_strategy = 'merge',
+	unique_key = ['block_date', 'cover_block_number', 'statusNum', 'evt_tx_hash', 'evt_index'],
```
Hi @sharkxff , the unique_key list here should match the columns in uniqueness test in schema.yml

Hosuke commented:
Hi @sharkxff, as this is a new model, we may need to have some more tests on it.
You can add some test cases(which have been manually verified on etherscan.io) in the seed file.
The seed file directory would be like `seeds/nxm/ethereum/nxm_ethereum_trades_seeds.csv`.
You can check these example seeds for reference: [perpetual_trades_seeds](https://github.com/duneanalytics/spellbook/blob/main/seeds/perpetual/trades/perpetual_trades_seed.csv), [x2y2_seeds](https://github.com/duneanalytics/spellbook/blob/main/seeds/x2y2/ethereum/x2y2_ethereum_trades_etherscan.csv), [dex_trade_seeds](https://github.com/duneanalytics/spellbook/blob/main/seeds/dex/trades/dex_trades_seed.csv).
Then we can use these seeds to write more specific tests in the schema.yml.

dune-eng commented:
Workflow run id 3392381793 approved.

dune-eng commented:
Workflow run id 3392381790 approved.

Hosuke commented:
If you can add some tests on specific columns using [generic test](https://github.com/duneanalytics/spellbook/tree/main/tests/generic) that would be better.

dune-eng commented:
Workflow run id 3425212183 approved.

dune-eng commented:
Workflow run id 3425212184 approved.

dune-eng commented:
Workflow run id 3425836271 approved.

dune-eng commented:
Workflow run id 3425836268 approved.

Hosuke commented on dbt_project.yml:
```diff
@@ -582,6 +589,16 @@ seeds:
         +column_types:
           master_address: string
           alternative_address: string
+    nxm:
+      +enabled: true
+      +schema: test_data
+      ethererum:
+        nxm_ethereum_trades_seed:
+          +column_types:
+            evt_tx_hash: string
+            block_date: date
+            #TODO Please fill in column name and its type for the seed
+
```
Hi @sharkxff , please follow the pattern and fill in the data type for the columns in the seed file.

Hosuke commented on models/nxm/ethereum/nxm_ethereum_schema.yml:
```diff
@@ -0,0 +1,120 @@
+      - &curr
+        name: curr
+        description: "curr"
+      - &premium
+        name: premium
+        description: "premium"
+        tests:
+          - nxm_trades_column_values:
+              seed_file_location: ref('nxm_ethereum_trades_seed')
```
And we can test specific column values like this way

Hosuke reviewed (dismissed):
Hi @sharkxff , please make the seed available for the project by adding it to dbt_project.yml

dune-eng commented:
Workflow run id 3433583829 approved.

dune-eng commented:
Workflow run id 3435037326 approved.

dune-eng commented:
Workflow run id 3435037319 approved.

dune-eng commented:
Workflow run id 3435126510 approved.

dune-eng commented:
Workflow run id 3435126501 approved.

dune-eng commented:
Workflow run id 3435159849 approved.

dune-eng commented:
Workflow run id 3435159832 approved.

dune-eng commented:
Workflow run id 3435297124 approved.

dune-eng commented:
Workflow run id 3435297136 approved.

Hosuke commented on dbt_project.yml:
```diff
@@ -469,6 +469,13 @@ models:
         +schema: hashflow_ethereum
         +materialized: view
 
+    nomad:
+      +schema: nexusmutual
+      +materialized: view
+      ethereum:
+        +schema: nexusmutual_ethereum
+        +materialized: view
+
```
You may mistakenly name the model as `nomad`, which is another existing model. I will commit a fix.

Hosuke commented:
To whom it may concern, I have reviewed this PR and here is the summary:

- Metadata looks good ✅
- Some column names in the schema file are not lowercased, but match the names in the source decoded table.
- Config blocks match uniqueness tests in schema files ✅
- Tests and seeds added for the new models ✅
- Incremental filters added when joining large tables ✅
- Contract addresses are lowercased ✅
- Inner join to large tables(ethereum.transactions) as possible ✅

jeff-dude commented:
thank you @Hosuke & @sharkxff -- below are my notes:
- renamed anything ```nxm``` to ```nexusmutual``` to match standard at decoded table level, let's strive for consistency
- moved forward with column rename to lowercase/underscore, i see hosuke mentioned, but i think we should follow through
- moved the date filters to within join conditions themselves for consistent approach across spellbook
- added blockchain filter on join to erc20 tokens
- removed descriptions in schema file which didn't add value past column name. they aren't a must if we don't have good descriptions

with the above changes, i didn't make huge logic changes outside of the blockchain filter on join to erc20 tokens (which also is an inner join, usually we do left join since it's manually maintained and can get behind -- should this update?). however, the gh action is now failing on duplicates post-incremental run. can someone validate the dupes and see why?

Hosuke commented:
> with the above changes, i didn't make huge logic changes outside of the blockchain filter on join to erc20 tokens (which also is an inner join, usually we do left join since it's manually maintained and can get behind -- should this update?).

I also noticed is the inner join with erc20 tokens, maybe nexusmutual only supports limited erc20 tokens and all listed in the table. I think we need @sharkxff to confirm this.

dune-eng commented:
Workflow run id 3442029609 approved.

dune-eng commented:
Workflow run id 3442029611 approved.

dune-eng commented:
Workflow run id 3442287689 approved.

dune-eng commented:
Workflow run id 3442287688 approved.

dune-eng commented:
Workflow run id 3442415377 approved.

dune-eng commented:
Workflow run id 3442415380 approved.

sharkxff commented:
After the compiled sql runs group by unique_key(block_date,cover_block_number,status_num,evt_tx_hash,evt_index ) in the DUNE UI, the result is" No results from query". However, dbt-test shows that !=0 unique_key results can be found. This part needs help to check.

sharkxff commented:
![image](https://user-images.githubusercontent.com/45616830/201262224-808ac821-6e1c-467c-8579-9e45d2f7e0a2.png)

sharkxff commented:
![image](https://user-images.githubusercontent.com/45616830/201262407-16a86187-5273-4a64-b0d3-31bbf9b792c5.png)

Hosuke commented on models/nexusmutual/ethereum/nexusmutual_ethereum_quotation_trades.sql:
```diff
@@ -0,0 +1,99 @@
+    {% endif %}
+),
+quo_cse_data as (
+    select * FROM (
+        select  *,
+            row_number() over(partition by cid order by evt_block_time desc) as time_rank
+        FROM {{ source('nexusmutual_ethereum', 'QuotationData_evt_CoverStatusEvent') }}
+    ) as quo_cse
```
Please try to avoid using `select *` for performance gain, unless all columns needed(In that case you still can list out all columns and make it clearer for review and maintainance).

Hosuke commented:
> ![image](https://user-images.githubusercontent.com/45616830/201262407-16a86187-5273-4a64-b0d3-31bbf9b792c5.png)

I write the query to show out [all grouped results with count(*)](https://dune.com/queries/1584259?d=1). It seems each row appears only 1 time. I will have a closer look for this issue.

Hosuke commented:
I think the [test failure] is caused by the unique keys is not strong enough for the incremental merge, which means there are overlapping part during the incremental update.

sharkxff commented:
But before yesterday, the unique value judgment was success

Hosuke commented:
> But before yesterday, the unique value judgment was success

Can you add more descriptions about columns in the schema.yml and sources.yml files for the model? As we need to understand this model better to help to solve the issue.

dune-eng commented:
Workflow run id 3443822556 approved.

dune-eng commented:
Workflow run id 3443822561 approved.

Hosuke commented:
Wait, the checks passed!

sharkxff commented:
> > But before yesterday, the unique value judgment was success
> 
> Can you add more descriptions about columns in the schema.yml and sources.yml files for the model? As we need to understand this model better to help to solve the issue.

I changed the table structure and removed some fields, and the compile has been successful so far

jeff-dude reviewed (approved):
thank you @sharkxff and @Hosuke 🤝 
i pushed one last fix for seed file. the join wasn't returning any rows, giving us false positive test results, due to bad seed data.
